File: go/hashmap/380_randomized_set.go
package hashmap

import "math/rand"

// O(1) 时间插入、删除和获取随机元素
// O(1)获取随机元素  边长数组可以实现
// O(1) 插入、删除 hashmap 可以实现
type RandomizedSet struct {
	data []int       // 存储元素
	idx  map[int]int // 存储每个元素在切片中的下标 {val： idx}
}

func ConstructorR() RandomizedSet {
	data := make([]int, 0)
	idx := make(map[int]int)
	return RandomizedSet{
		data: data,
		idx:  idx,
	}
}

// Insert 插入一个元素
func (r *RandomizedSet) Insert(val int) bool {
	if _, ok := r.idx[val]; ok {
		return false
	}
	// 先存储下标
	r.idx[val] = len(r.data)
	// 数据
	r.data = append(r.data, val)
	return true
}

// Remove 删除元素时需要注意O(1)的话 要把需要删除的元素移到最后面
func (r *RandomizedSet) Remove(val int) bool {
	if idx, ok := r.idx[val]; ok {

		// 在data 中的索引下标是 idx
		last := len(r.data) - 1

		// first update idx
		r.idx[r.data[last]] = idx

		// second swap data
		r.data[idx] = r.data[last]

		delete(r.idx, val)
		// third, remove last
		r.data = r.data[:last]
		return true
	}

	return false
}

func (r *RandomizedSet) GetRandom() int {
	// return r.data[rand.Intn(len(r.data))]
	randIdx := rand.Intn(len(r.data))
	return r.data[randIdx]
}

File: rust/hashmap/380_randomized_set.rs


use std::collections::HashMap;
use rand::Rng;
struct RandomizedSet {
    data: Vec<i32>,
    index: HashMap<i32, usize>,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl RandomizedSet {

    fn new() -> Self {
        let data = vec![];
        let index = HashMap::new();

        RandomizedSet{
            data,
            index,
        }
    }
    
    fn insert(&mut self, val: i32) -> bool {
        if self.index.contains_key(&val) {
            return false
        }

        // data length
        let length = self.data.len();
        // store index
        self.index.insert(val, length);
        // store data
        self.data.push(val);

        true
    }
    
    fn remove(&mut self, val: i32) -> bool {
        if !self.index.contains_key(&val) {
            return false
        }

        // 和最后一个元素交换后再删除
        let last = self.data.pop().unwrap();
        let idx = self.index.remove(&val).unwrap();
        // FIX 最后一个元素不需要交换更新，直接删除即可
        if idx != self.data.len() {
            // index update
            self.index.insert(last, idx);
            // 数据udpate
            self.data[idx] = last;
        }

        true
    }
    
    fn get_random(&self) -> i32 {

        let mut rng = rand::thread_rng();
        let idx = rng.gen_range(0..self.data.len());

        self.data[idx]
    }
}